resolver/dns/go19.go

```diff
@@ -0,0 +1,87 @@
+// +build go1.9
+
+/*
+ *
+ * Copyright 2018 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package dns
+
+import (
+	"fmt"
+	"net"
+	"strings"
+
+	"golang.org/x/net/context"
+)
+
+var customAuthorityDialler = func(authority string) func(ctx context.Context, network, address string) (net.Conn, error) {
+	return func(ctx context.Context, network, address string) (net.Conn, error) {
+		var dialer net.Dialer
+		return dialer.DialContext(ctx, network, authority)
+	}
+}
+
+func ensureAuthorityPort(target string) (targetWithPort string, err error) {
+
+	var host string
+	var port string
+
+	if host, port, err = net.SplitHostPort(target); err == nil {
+		if port == "" {
+			// If the port field is empty (target ends with colon), e.g. "[::1]:", this is an error.
+			return "", errEndsWithColon
+		}
+		// target has port, i.e ipv4-host:port, [ipv6-host]:port, host-name:port
+		if host == "" {
+			// Keep consistent with net.Dial(): If the host is empty, as in ":80", the local system is assumed.
+			host = "localhost"
+		}
+	} else {
+		// target could be ipv4-host, [ipv6-host], or host-name without port
+		if host, port, err = net.SplitHostPort(fmt.Sprintf("%s:53", target)); err != nil {
+			if strings.Contains(target, ":") {
+				// target could be ipv6-host without port
+				if host, port, err = net.SplitHostPort(fmt.Sprintf("[%s]:53", target)); err != nil {
+					return "", fmt.Errorf("invalid authority address %v", target)
+				}
+			} else {
+				return "", fmt.Errorf("invalid authority address %v", target)
+			}
+		}
+	}
+
+	if strings.Contains(host, ":") {
+		// target host is ipv6
+		return fmt.Sprintf("[%s]:%s", host, port), nil
+	}
+
+	// target host is ipv4 or host-name
+	return fmt.Sprintf("%s:%s", host, port), nil
```

Comment:
It seems `ensureAuthorityPort` share substantial logic with `parseTarget`. Could you please merge these two functions into one. So `parseTarget` takes an extra argument like `defaultPort`. And here for `ensureAuthorityPort`, you calls `parseTarget` with the new arg to be "53", and then call the `net.JoinHostPort` on the returned host and port.